src/WinRT.Runtime2/InteropServices/Exceptions/IRestrictedErrorInfoMethods.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Runtime.InteropServices;
+
+namespace WindowsRuntime.InteropServices;
+
+/// <summary>
+/// Provides low-level helpers to call <c>IRestrictedErrorInfo</c> methods through a vtable pointer,
+/// and to safely marshal returned BSTR values to managed strings.
+/// </summary>
+/// <remarks>
+/// These helpers assume "thisPtr" param points to a valid COM object implementing
+/// the <c>IRestrictedErrorInfo</c> interface and that HRESULTs are validated via the
+/// <c>Assert()</c> helper on the returned status code. Any BSTRs returned by the COM methods
+/// are converted to <see cref="string"/> and then released with <see cref="Marshal.FreeBSTR(System.IntPtr)"/>.
+/// </remarks>
+internal static class IRestrictedErrorInfoMethods
+{
+    /// <summary>
+    /// Retrieves error details from an <c>IRestrictedErrorInfo</c> instance.
+    /// </summary>
+    /// <param name="thisPtr">A raw pointer to the COM object implementing <c>IRestrictedErrorInfo</c>.</param>
+    /// <param name="description">On return, receives the human-readable error description (may be empty).</param>
+    /// <param name="error">On return, receives the HRESULT error code.</param>
+    /// <param name="restrictedDescription">On return, receives the restricted error description (may be empty).</param>
+    /// <param name="capabilitySid">On return, receives the capability SID (may be empty).</param>
+    /// <remarks>
+    /// All returned BSTRs are freed regardless of success or failure. Empty strings are returned if
+    /// any corresponding native pointer is null.
+    /// </remarks>
+    public static unsafe void GetErrorDetails(
+        void* thisPtr,
+        out string description,
+        out int error,
+        out string restrictedDescription,
+        out string capabilitySid)
+    {
+        void* _description = null;
+        void* _restrictedDescription = null;
+        void* _capabilitySid = null;
+
+        try
+        {
+            fixed (int* pError = &error)
+            {
+                ((IRestrictedErrorInfoVftbl*)*(void***)thisPtr)->GetErrorDetails(
+                    thisPtr,
+                    &_description,
+                    pError,
+                    &_restrictedDescription,
+                    &_capabilitySid).Assert();
+            }
+
+            description = _description != null ? Marshal.PtrToStringBSTR((nint)_description) : string.Empty;
+            restrictedDescription = _restrictedDescription != null ? Marshal.PtrToStringBSTR((nint)_restrictedDescription) : string.Empty;
+            capabilitySid = _capabilitySid != null ? Marshal.PtrToStringBSTR((nint)_capabilitySid) : string.Empty;
```

Comment:
We can avoid this dance and make this simpler by using the new marshaller, same below too to free:
```csharp
description = BStrStringMarshaller.ConvertToManaged(_description);
```
We'll need to make the out parameters nullable as well.